blog/2019-02-19-reliable-microservices-data-exchange-with-the-outbox-pattern.adoc

```diff
@@ -0,0 +1,598 @@
+When receiving a delete event (`op` = `d`), it will discard that event,
+as that deletion of event records from the outbox table is only a technicality not relevant to downstream consumers.
+Things get more interesting, when receiving a create event (`op` = `c`).
+Such record will be propagated to Apache Kafka.
+
+Debezium's change events have a complex structure, that contain the old (`before`) and new (`after`) state of the represented row.
+The event structure to propagate is obtained from the `after` state.
+The `aggregatetype` value from the captured event record is used to build the name of the topic to send the event to.
+For instance, events with `aggregatetype` set to `Order` will be sent to the `OrderEvents` topic.
+`aggregateid` is used as the message key, making sure all messages of that aggregate will go into the same partition of that topic.
+The message value is a structure comprising the original event payload (encoded as JSON),
+the timestamp indicating when the event was produced and the event type.
+Finally, the event UUID is propagated as a Kafka header field.
+This allows for efficient duplicate detection by consumers, without having to examine the actual message contents.
+
+== Events in Apache Kafka
+
+Now let's take a look into the `OrderEvents` and `CustomerEvents` topics.
+
+If you have checked out the example sources and started all the components via Docker Compose
+(see the _README.md_ file in the example project for more details),
+you can place purchase orders via the order service's REST API like so:
+
+[source]
+----
+cat resources/data/create-order-request.json | http POST http://localhost:8080/order-service/rest/orders
+----
+
+Similarly, specific order lines can be canceled:
+
+[source]
+----
+cat resources/data/cancel-order-line-request.json | http PUT http://localhost:8080/order-service/rest/orders/1/lines/2
+----
+
+When using a tool such as the very practical https://github.com/edenhill/kafkacat[kafkacat] utility,
+you should now see messages like these in the `OrderEvents` topic:
+
+[source]
+----
+kafkacat -b kafka:9092 -C -o beginning -f 'Headers: %h\nKey: %k\nValue: %s\n' -q -t OrderEvents
+----
+
+[source]
+----
+Headers: eventId=d03dfb18-8af8-464d-890b-09eb8b2dbbdd
+Key: "4"
+Value: {"eventType":"OrderCreated","ts_ms":1550307598558,"payload":"{\"id\": 4, \"lineItems\": [{\"id\": 7, \"item\": \"Debezium in Action\", \"status\": \"ENTERED\", \"quantity\": 2, \"totalPrice\": 39.98}, {\"id\": 8, \"item\": \"Debezium for Dummies\", \"status\": \"ENTERED\", \"quantity\": 1, \"totalPrice\": 29.99}], \"orderDate\": \"2019-01-31T12:13:01\", \"customerId\": 123}"}
+Headers: eventId=49f89ea0-b344-421f-b66f-c635d212f72c
+Key: "4"
+Value: {"eventType":"OrderLineUpdated","ts_ms":1550308226963,"payload":"{\"orderId\": 4, \"newStatus\": \"CANCELLED\", \"oldStatus\": \"ENTERED\", \"orderLineId\": 7}"}
+----
+
+The `payload` field with the message values is the string-ified JSON representation of the original events.
```

Comment:
The article does not explain how the sringification was achieved.

Reply:
Clarified it.